ext/url/01_urlpattern.js

```diff
@@ -55,10 +58,68 @@ const COMPONENTS_KEYS = [
  * @property {string[]} groupNameList
  */
 
+/**
+ * @template K
+ * @template V
+ */
+class LRUCache {
+  /** @type {Map<K, V>} */
+  #map = new SafeMap();
+  #capacity = 0;
+
+  /** @type {K} */
+  #lastUsedKey = undefined;
+  /** @type {V} */
+  #lastUsedValue = undefined;
+
+  /** @param {number} capacity */
+  constructor(capacity) {
+    this.#capacity = capacity;
+  }
+
+  /**
+   * @param {K} key
+   * @param {(key: K) => V} factory
+   * @return {V}
+   */
+  getOrInsert(key, factory) {
+    if (this.#lastUsedKey === key) return this.#lastUsedValue;
+    const value = this.#map.get(key);
+    if (value !== undefined) {
+      if (MathRandom() < 0.1) {
```

Comment:
Ideally should be actually do a least recently used strategy but fine for now. Maybe rename class to `RandomLruCache`

Reply:
It does do least recently used - maps are insertion ordered. See the code below this line (the re-insertion is to put the used item back to the "hot" side of the LRU), we use `keys().next().value` to get item at the "cold" end of the LRU.

I am intentionally using a random chance to determine whether to refresh / insert as a very cheap pseudo "young space" where if a item is not used at least 10 times on average (but not 10 times in a row), it doesn't graduate into the more permanent LRU storage. This is done because inserting and refreshing LRU items is actually comparatively very expensive (about 50% of the operation it is caching).

This is a "pseudo young space" because it is random based, and thus sometimes misses the mark. In the benchmark configurations I have run that were written based on usage experience from real websites using Fresh (like dotcom), where there are some very hot routes, and a long tail of cold routes, I think this will do well.

Reply:
But yes, I can rename to `SampledLruCache`

Reply:
Sounds good, thanks for explaning